test/unit/utils/OrderedMsgChain.test.ts

```diff
@@ -335,23 +455,103 @@ describe('OrderedMsgChain', () => {
         }, () => {}, 50)
         util.add(msg1)
         shuffled.forEach((msg) => {
-            if (msg.getTimestamp() !== msg1.getTimestamp()) {
-                util.add(msg)
-            }
+            util.add(msg)
         })
+
         try {
             assert.deepStrictEqual(received, expected)
         } catch (e) {
-            const shuffledTimestamps: number[] = []
-            shuffled.forEach((streamMessage: StreamMessage) => {
-                shuffledTimestamps.push(streamMessage.getTimestamp())
+            const timestamps: number[] = []
+            expected.forEach((streamMessage: StreamMessage) => {
+                timestamps.push(streamMessage.getTimestamp())
             })
             const receivedTimestamps: number[] = []
             received.forEach((streamMessage: StreamMessage) => {
                 receivedTimestamps.push(streamMessage.getTimestamp())
             })
             throw new Error('Was expecting to receive messages ordered per timestamp but instead received timestamps in this '
-                + `order:\n${receivedTimestamps}.\nThe unordered messages were processed in the following timestamp order:\n${shuffledTimestamps}`)
+                + `order:\n${receivedTimestamps}.\nThe unordered messages were processed in the following timestamp order:\n${timestamps}`)
         }
     })
+
+    it('handles unordered messages in order with gapfill (large randomized test)', (done) => {
+        // this test breaks a large number of messages in random order, with duplicates, into chunks
+        // each time queue is drained or gap is detected, it adds the next chunk of messages.
+        const expected = [msg1]
+        const NUM_CHUNKS = 12
+        for (let i = 2; i <= 1000; i++) {
+            expected.push(createMsg({
+                timestamp: i,
+                sequenceNumber: 0,
+                prevTimestamp: i - 1,
+                prevSequenceNumber: 0
+            }))
+        }
+        const duplicates = shuffle(expected).slice(0, expected.length / 3)
+        const shuffled = shuffle([...duplicates, ...expected])
+        const chunks = chunk(shuffled, Math.floor(expected.length / NUM_CHUNKS))
+        function next() {
+            const result = nextChunk()
+            util.debugStatus()
+            if (result) {
+                return
+            }
+            setTimeout(() => {
+                checkDone()
+            }, 0)
+        }
+
+        function nextChunk() {
+            const items = chunks.pop()
+            if (!items) { return false }
+            items.forEach((msg) => {
+                util.add(msg)
+            })
+            return true
+        }
+
+        function checkDone() {
+            clearTimeout(t)
```

Comment:
Reading from top to bottom, got a bit confused about what `t` is since it is declared later in the scope.

Reply:
Good point

Reply:
Fixed.